video_creation/final_video.py

```diff
@@ -3,159 +3,278 @@
-from moviepy.audio.io.AudioFileClip import AudioFileClip
-from moviepy.video.VideoClip import ImageClip
-from moviepy.video.compositing.CompositeVideoClip import CompositeVideoClip
-from moviepy.video.compositing.concatenate import concatenate_videoclips
-from moviepy.video.io.VideoFileClip import VideoFileClip
-from moviepy.video.io.ffmpeg_tools import ffmpeg_extract_subclip
+from typing import Tuple, Any, Union
+
+from moviepy.editor import (
+    VideoFileClip,
+    AudioFileClip,
+    ImageClip,
+    CompositeAudioClip,
+    CompositeVideoClip,
+)
 from rich.console import Console
+from rich.progress import track
+from attr import attrs
 
 from utils.cleanup import cleanup
 from utils.console import print_step, print_substep
 from utils.video import Video
 from utils.videos import save_data
 from utils import settings
+from video_creation.background import download_background, chop_background_video
+
+
+@attrs
+class FinalVideo:
+    video_duration: int = 0
+    console = Console()
+
+    def __attrs_post_init__(self):
+        self.W: int = int(settings.config["settings"]["video_width"])
+        self.H: int = int(settings.config["settings"]["video_height"])
+
+        if not self.W or not self.H:
+            self.W, self.H = 1080, 1920
 
-console = Console()
-W, H = 1080, 1920
+        self.vertical_video: bool = self.W < self.H
 
+        self.max_length: int = int(settings.config["settings"]["video_length"])
+        self.time_before_first_picture: float = settings.config["settings"]["time_before_first_picture"]
+        self.time_before_tts: float = settings.config["settings"]["time_before_tts"]
+        self.time_between_pictures: float = settings.config["settings"]["time_between_pictures"]
+        self.delay_before_end: float = settings.config["settings"]["delay_before_end"]
 
-def name_normalize(name: str) -> str:
-    name = re.sub(r'[?\\"%*:|<>]', "", name)
-    name = re.sub(r"( [w,W]\s?\/\s?[o,O,0])", r" without", name)
-    name = re.sub(r"( [w,W]\s?\/)", r" with", name)
-    name = re.sub(r"(\d+)\s?\/\s?(\d+)", r"\1 of \2", name)
-    name = re.sub(r"(\w+)\s?\/\s?(\w+)", r"\1 or \2", name)
-    name = re.sub(r"\/", r"", name)
-    name[:30]
+        self.opacity = settings.config["settings"]["opacity"]
+        self.opacity = 1 if self.opacity is None or self.opacity >= 1 else self.opacity
 
-    lang = settings.config["reddit"]["thread"]["post_lang"]
-    if lang:
-        import translators as ts
+    @staticmethod
+    def name_normalize(
+            name: str
+    ) -> str:
+        name = re.sub(r'[?\\"%*:|<>]', "", name)
+        name = re.sub(r"( [w,W]\s?/\s?[oO0])", r" without", name)
+        name = re.sub(r"( [w,W]\s?/)", r" with", name)
+        name = re.sub(r"(\d+)\s?/\s?(\d+)", r"\1 of \2", name)
+        name = re.sub(r"(\w+)\s?/\s?(\w+)", r"\1 or \2", name)
+        name = re.sub(r"/", "", name)
+        # name[:30]  # does nothing
```

Comment:
It shortens the file name because some OS's have relatively short limits to how long a file name can be